data_sync/mvds.md

```diff
@@ -0,0 +1,130 @@
+# Minimum Viable Data Synchronization
+
+**DRAFT VERSION 0.5**
+
+*Written by Oskar Thorén oskar@status.im & Dean Eigenmann dean@status.im*
+
+## Table of Contents
+
+1. [Abstract](#abstract)
+2. [Definitions](#definitions)
+3. [Wire Protocol](#wire-protocol)
+    1. [Secure Transport](#secure-transport) 
+    2. [Payloads](#payloads)
+4. [Synchronization](#synchronization)
+    1. [State](#state)
+    2. [Flow](#flow)
+    3. [Retransmission](#retransmission)
+5. [Footnotes](#footnotes)
+
+## Abstract
+
+In this specification we describe a minimum viable protocol for data synchronization inspired by the Bramble Synchronization Protocol<sup>1</sup>. This protocol is designed to ensure reliable messaging between peers communicating across an unreliable peer-to-peer (P2P) network where they may be unreachable or unresponsive.
+
+We present a functional specification for future implementation<sup>2</sup> as well as reference simulation data which demonstrates its performance.
+
+## Definitions
+
+| Term | Description |
+|------|-------------|
+| **Peer** | The other nodes that a node is connected to. |
+| **Record** | Defines a payload element of either the type `OFFER`, `REQUEST`, `MESSAGE` or `ACK` |
+
+## Wire Protocol
+
+### Secure Transport
+
+This specification does not define anything related to the transport of packets. It is assumed that this is abstracted in such a way that any secure transport protocol could be easily implemented. Likewise, properties such as confidentiality, integrity, authenticity and forward secrecy are assumped to be provided by a layer below.
+
+### Payloads
+
+Payloads are implemented using [protocol buffers](https://developers.google.com/protocol-buffers/).
+
+```protobuf
+message Payload {
+  Ack ack = 1;
+  Offer offer = 2;
+  Request request = 3;
+  repeated Message messages = 4;
+}
+
+message Ack {
+  repeated bytes id = 1;
+}
+
+message Message {
+  bytes group_id = 1;
+  int64 timestamp = 2;
+  bytes body = 3;
+}
+
+message Offer {
+  repeated bytes id = 1;
+}
+
+message Request {
+  repeated bytes id = 1;
+}
+```
+
+## Synchronization
+
+### State
+
+State is kept for any record of the types `OFFER`, `REQUEST` and `MESSAGE` we do not keep states for `ACK` records as we do not retransmit those periodically. State is stored per peer. The following information is stored for records:
```

Comment:
Can you elaborate on these record types? Does the sender resend messages that were not responded with an ACK? What is the purpose of ACK other than to know that the recipient received it? 

Can message delivery / retry be done on the transport layer?

Reply:
@prestonvanloon is this not clear enough in the flow section?

Reply:
> Can message delivery / retry be done on the transport layer?

Not really, that makes assumptions on the lower level transport.

Reply:
It's clear now with the new protobuf payload message. Thanks!